add parser support for oneof and map fields.

AlmogCohen opened the pull request:
I'm mostly interested in the parser and not in the C extensions generated so I've improved the parser while keeping its default behaviour intact.
I'm not a C guy so nothing I can help with adding the complete support for the new fields. I do hope it might help folks down the line add the full support.

Commit includes:
* add support for oneof field in parser.
* add support for map fields in parser.
* add capability to disable specific tokens in parser.
* disable the new oneof and map field by default in the parser to make sure the parser works with the C extensions out-of-the-box.
* add informative exceptions for disabled parser tokens.
* simplify the `tokenize` code by creating parsable tokens mapping from name to class.
* change a few function-level variable names to improve clarity

AlmogCohen commented:
I couldn't get the tests to work locally following the instructions as all unit tests error out. I'm convinced that none of the tests were broken as I've changed nothing in the existing parser's design. I've also tested parsing of some files I work with before and after the change. I'd be happy to run the tests again if proper instructions will be given. The errors look like that:
```
    def test_sint64(proto_lib):
>       from test_signed_integer_proto import Int
E       ModuleNotFoundError: No module named 'test_signed_integer_proto'
```

AlmogCohen commented:
Might be relevant for https://github.com/appnexus/pyrobuf/issues/109 and https://github.com/appnexus/pyrobuf/issues/95 if someone would like to implement the C side of `map` and `oneof`.

tburmeister commented on pyrobuf/parse_proto.py:
```diff
@@ -22,6 +22,8 @@ class Parser(object):
         ('MODIFIER', r'(optional|required|repeated)'),
         ('FIELD', r'([A-Za-z][0-9A-Za-z_]*)\s+([A-Za-z][0-9A-Za-z_]*)\s*='
                   r'\s*(\d+)'),
+        ('MAP_FIELD', r'map<([A-Za-z][0-9A-Za-z_]+),\s*([A-Za-z][0-9A-Za-z_]+)>\s+([A-Za-z][0-9A-Za-z_]*)\s*='
```
This is nit-picky, but can you move more of the first line to the second line and align the second line after the comma?

AlmogCohen replied:
I'm OK with that...what is the general line length you follow? 80? We can add `flake8` and add the line length configuration to the `tox.ini` so there won't be any guess work for people in next PRs. What do you think?

tburmeister replied:
I've been trying to stick to 80. I think adding `flake8` and a `tox.ini` is a great idea!

tburmeister commented on pyrobuf/parse_proto.py:
```diff
@@ -505,49 +558,73 @@ def add_cython_info(self, message):
         type = None
 
     class Option(Token):
+        token_type = 'OPTION'
```
I like this change!

tburmeister commented on pyrobuf/parse_proto.py:
```diff
@@ -252,25 +234,34 @@ def parse(self, cython_info=True, fname='', includes=None):
+            cls,
+            fname,
+            includes,
+            disabled_tokens=tokens_not_supported_by_c
```
Think this needs to be `self.tokens_not_supported_by_c`?

AlmogCohen replied:
Since it is within the function definition (the args) it is considered to be the still in the class's namespace. Nor the `self` nor `cls` are needed in this level. And unittests pass that way anyway ;)

tburmeister commented on pyrobuf/parse_proto.py:
```diff
@@ -41,6 +43,34 @@ class Parser(object):
+        'EXTEND',
+    )
+
+    tokens_not_supported_by_c = (
```
Can we change this to just `unsupported_tokens`?

AlmogCohen replied:
`unsupported_tokens` is wiser, I just wanted to somehow communicate that it is supported by the parser, but not by the code that make it into a C extensions later on...Do you think there would be a good way to convey that?

AlmogCohen replied:
I take your input and put for some documentation if someone try to dive deeper

tburmeister commented on pyrobuf/parse_proto.py:
```diff
@@ -124,68 +156,17 @@ def tokenize(self):
-
-            elif token_type == 'DEFAULT':
-                yield self.Default(line, *vals)
+            assert token_type not in disabled_token_types, (
```
I think we may just want to throw a warning here, or make failing or warning an option.

AlmogCohen replied:
Currently the map field and oneof where just failing the code without good explanation. I do worry that if we go for warnings and skip the assertions, people we'll get weird exceptions in the code that make the C extensions and it would be hard to trace the origin.

Since there will be exceptions down the line anyway, I preferred to raise them where it would make sense and the user would sure understand why things are not working. What do you think?

tburmeister replied:
Okay, that makes sense.

tburmeister commented on pyrobuf/parse_proto.py:
```diff
@@ -252,25 +234,34 @@ def parse(self, cython_info=True, fname='', includes=None):
 
     @classmethod
-    def parse_from_filename(cls, fname, includes):
+    def parse_from_filename(
```
I don't love this style for function declarations, but I guess it's fine.

AlmogCohen replied:
I did the line breaks because it was over the 80 line limit I saw the file written with. Anyway, after renaming the variable it is now a one line definition again :)

tburmeister commented:
> P.S. It would be nice to have the C side of the new fields implemented even though I'm not in need for the speed. I'm only looking for a convenient API for working with `.proto` definitions in python. I'm thinking maybe writing something to translate `.proto` file into [schematics](https://schematics.readthedocs.io/en/latest/) python classes.

Agreed - hopefully I will get to it someday but realistically it won't happen until I or someone willing to implement it needs the functionality.

tburmeister commented:
> I couldn't get the tests to work locally following the instructions as all unit tests error out. I'm convinced that none of the tests were broken as I've changed nothing in the existing parser's design. I've also tested parsing of some files I work with before and after the change. I'd be happy to run the tests again if proper instructions will be given. The errors look like that:
> 
> ```
>     def test_sint64(proto_lib):
> >       from test_signed_integer_proto import Int
> E       ModuleNotFoundError: No module named 'test_signed_integer_proto'
> ```

It's trying to import a built protobuf spec - you first need to build and install all the test message specs.

tburmeister commented:
> > I couldn't get the tests to work locally following the instructions as all unit tests error out. I'm convinced that none of the tests were broken as I've changed nothing in the existing parser's design. I've also tested parsing of some files I work with before and after the change. I'd be happy to run the tests again if proper instructions will be given. The errors look like that:
> > ```
> >     def test_sint64(proto_lib):
> > >       from test_signed_integer_proto import Int
> > E       ModuleNotFoundError: No module named 'test_signed_integer_proto'
> > ```
> 
> It's trying to import a built protobuf spec - you first need to build and install all the test message specs.

Try running `pyrobuf --install tests/proto/` before running the tests. The instructions for testing are not clear and testing should be simplified - I will try to make that happen - thanks for bringing this to light.

tburmeister commented:
It would be great if we could add a test message spec to demonstrate that this works.

tburmeister commented:
I checked out your branch locally and all tests pass for me.

AlmogCohen commented:
> > > I couldn't get the tests to work locally following the instructions as all unit tests error out. I'm convinced that none of the tests were broken as I've changed nothing in the existing parser's design. I've also tested parsing of some files I work with before and after the change. I'd be happy to run the tests again if proper instructions will be given. The errors look like that:
> > > ```
> > >     def test_sint64(proto_lib):
> > > >       from test_signed_integer_proto import Int
> > > E       ModuleNotFoundError: No module named 'test_signed_integer_proto'
> > > ```
> > 
> > 
> > It's trying to import a built protobuf spec - you first need to build and install all the test message specs.
> 
> Try running `pyrobuf --install tests/proto/` before running the tests. The instructions for testing are not clear and testing should be simplified - I will try to make that happen - thanks for bringing this to light.

I've added that to the `README.md` and created a small `TESTING` section there. Now the tests do run locally :)

AlmogCohen commented:
> It would be great if we could add a test message spec to demonstrate that this works.

I'll write some tests now. The only thing that bothers me that the tests I'm about to write will care only about the parser and would not be similar to the rest of the tests that actually build the `proto` files. Would it be OK to create a  `test_parser` file and nest multiple unit tests there? I can then parse text from strings within the file and not a whole `proto` file as I don't need it to be built. Does that sound right?

AlmogCohen commented:
I think it would actually be better to create a `test_parser` **folder** and nest some tests under that folder to create logical separation. A separation between tests that actually use the built objects with pyrobuf and tests that simply test the python parser code.

AlmogCohen commented:
I've added decent unittests for both `oneof` and `MapField`. I've also grouped all parser unittests grouped under `test_parser` folder. One of the benefits of doing so is allowing someone to run the pure python unittests without the need to build the C extensions. (pytest allow folder-specifc runs)

On another note, following your instructions I've successfully run the unittests but the second run failed with another error. Not a big deal for this PR but it would be nice to fix the issue though.
```
ImportError while importing test module '/Users/almogcohen/Clients/algosec/code/pyrobuf/tests/test_unicode_strings.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_unicode_strings.py:5: in <module>
    from proto_lib_fixture import proto_lib
E   ModuleNotFoundError: No module named 'proto_lib_fixture'
```

Thank you much for the code review and quick comments!

tburmeister commented:
This is really great! Thanks so much for doing this! Let me know when you're ready and I can merge.

AlmogCohen commented:
`flake8` added, all warnings fixed in the file I worked on. Just wanted to say that in my projects I try to stick to line length of 120 and not 80. 80 feels to me like a screen size from another era ;)

With the new line-length enforcement there are a few more of the function definition style you didn't like but I don't know of any other way!

Ready to merge B-)

tburmeister commented:
Great! Thanks again! 120 probably makes more sense... Maybe I'll change it at some point.

AlmogCohen commented:
Awesome thank you so much! :)
